### ISSUE TITLE
How to get the transport plan for wasserstein distance?

### ISSUE THREAD

**Issue by heslowen** [jeanfeydy/geomloss#18]
Hello!
How to get the transport plan for wasserstein distance? For example:
```python
    import torch
    from geomloss import SamplesLoss  # See also ImagesLoss, VolumesLoss
    from tensorboardX import SummaryWriter
    # Create some large point clouds in 3D
    x = torch.randn(100, 90, requires_grad=True).cuda()
    y = torch.randn(100, 90).cuda()
    x_weight =torch.softmax(torch.randn(100, requires_grad=True).cuda(),dim=-1)
    y_weight = torch.softmax(torch.randn(100, requires_grad=True).cuda(),dim=-1)
    # Define a Sinkhorn (~Wasserstein) loss between sampled measures
    loss = SamplesLoss(loss="sinkhorn", p=2, blur=.05)
    L = loss(x_weight, x,y_weight, y)  # By default, use constant weights = 1/number of samples
    g_x_weight, = torch.autograd.grad(L, [x_weight])  # GeomLoss fully supports autograd!
    print(g_x_weight)

output:
    tensor([38.3446, 40.1507, 54.9848, 24.2403, 36.6134, 21.7442, 40.9704, 37.5197,
        36.8129, 30.8433, 41.8184, 37.8540, 29.5628, 32.1728, 35.4814, 34.3499,
        30.5784, 29.4510, 30.2563, 35.3756, 34.6008, 21.2113, 49.6258, 33.0905,
        26.7779, 39.9584, 27.6789, 33.0485, 35.1336, 31.8450, 33.2152, 49.8619,
        27.9988, 31.8357, 32.6381, 36.5462, 22.4289, 34.1696, 23.5495, 37.9015,
        37.1510, 30.8962, 42.6643, 41.9360, 32.8569, 52.4191, 48.5090, 32.2841,
        39.2593, 32.8378, 24.9272, 29.9081, 27.0655, 25.6891, 36.1147, 38.3208,
        43.5676, 40.3591, 32.9711, 22.7926, 33.5953, 26.1191, 36.2213, 26.2793,
        29.8464, 37.0326, 43.1056, 30.1160, 42.6618, 31.8065, 36.9675, 24.9423,
        47.4061, 27.5967, 22.2735, 25.6721, 36.1677, 33.0408, 28.5409, 32.0098,
        31.1463, 33.6331, 28.8633, 20.3415, 32.0224, 32.7113, 29.5103, 49.1259,
        38.1893, 29.8794, 41.5620, 31.8520, 30.0871, 29.5324, 17.6336, 38.2577,
        38.8874, 28.6118, 36.6356, 31.2856], device='cuda:0')

```
And I want to check the meanning of g_x_weight. I mean, when I backward the loss and update params, what's the change in the model (asume x and x_weight are outputs of the model)? After update, will both x and x_weight be close to y and y_weight next time?

**Comment by jeanfeydy** [comment]
Hi @heslowen ,

As discussed for instance [in this tutorial](https://www.kernel-operations.io/geomloss/_auto_examples/optimal_transport/plot_optimal_transport_labels.html), the (blurry) optimal transport plan "π_ij" is encoded in the dual vectors (*potentials*) computed by the GeomLoss solver. (See e.g. [Gabriel Peyré's book](https://arxiv.org/abs/1803.00567) or my PhD thesis, available in January, for more details.)

In your case, something along the lines of:

```python
N, M, D = x.shape[0], y.shape[0], x.shape(1)  # Number of points, dimension
p = 2, blur = .05
OT_solver = SamplesLoss(loss = "sinkhorn", p = p, blur = blur, 
                        debias = False, potentials = True)
F, G = OT_solver(x_weight, x, y_weight, y)  # Dual potentials

a_i, x_i = x_weight.view(N,1), x.view(N,1,D)
b_j, y_j = y_weight.view(1,M), y.view(1,M,D)
F_i, G_j = F.view(N,1), G.view(1,M)

C_ij = (1/p) * ((x_i - y_j)**p).sum(-1)  # (N,M) cost matrix
eps = blur**p  # temperature epsilon
P_ij = ((F_i + G_j - C_ij) / eps).exp() * (a_i * b_j)  # (N,M) transport plan
```
would do the trick. Of course, your transport plan here will be an `(N,M)` matrix: to scale things up with large point clouds, you may consider encoding it as a semi-symbolic [KeOps `LazyTensor`](http://www.kernel-operations.io/keops/index.html).

As for your second question: `g_x_weights` is the gradient of the "Sinkhorn" loss function with respect to the weights of your Dirac masses - no more, no less. If you use it for your updates, your model will progressively give more importance to the samples `x` which are close to the target measure encoded by `(y_weight, y)`.
However, if you actually want your samples `x` to move around, you will need to compute the gradient:
```python
g_x, = torch.autograd.grad(L, [x])
```
This is explained in detail in [these](https://www.kernel-operations.io/geomloss/_auto_examples/sinkhorn_multiscale/plot_epsilon_scaling.html) two [tutorials](https://www.kernel-operations.io/geomloss/_auto_examples/optimal_transport/plot_optimal_transport_2D.html).
Does this answer your questions?

Best regards,
Jean

**Comment by heslowen** [comment]
You have a clear answer ! Thank you so much~

**Comment by kheyer** [comment]
Quick question off this. How can I recover the transport cost from the coupling matrix? I'm familiar with the POT library where `cost = (coupling * cost_matrix).sum()`, but that doesn't seem to be the case with Geomloss. For example:

```
loss1 = SamplesLoss(loss='sinkhorn', backend='tensorized', p=2, blur=0.05, scaling=0.5, debias=False)
loss2 = SamplesLoss(loss='sinkhorn', backend='tensorized', p=2, blur=0.05, scaling=0.5, debias=False, potentials=True)

loss1(x, y)
> tensor(445.7838)

F, G = loss2(x,y)

a_i = loss2.generate_weights(x).view(-1,1)
x_i = x.unsqueeze(1)

b_i = loss2.generate_weights(y).view(1,-1)
y_i = y.unsqueeze(0)

F_i, G_j = F.view(-1,1), G.view(1,-1)
p = 2

C_ij = (1/p) * ((x_i - y_i)**p).sum(-1)

eps = 0.05*p

P_ij = ((F_i + G_j - C_ij) / eps).exp() * (a_i * b_i)

(C_ij*P_ij).sum()
> tensor(1.5955)
```

How can I recover the output of `loss1` using the outputs of `loss2`?

Thanks

**Comment by ZhangYuanhan-AI** [comment]
> 

I also face this issue. Do you know how to solve this problem?
